apps/wallet/src/ui/app/redux/slices/sui-objects/Coin.ts

```diff
@@ -81,26 +81,60 @@ export class Coin {
      *
      * @param signer A signer with connection to fullnode
      * @param coins A list of Coins owned by the signer with the same generic type(e.g., 0x2::Sui::Sui)
+     * @param gasCoins A list of Sui coins owned by the signer
      * @param amount The amount to be staked
      * @param validator The sui address of the chosen validator
      */
     public static async stakeCoin(
         signer: SignerWithProvider,
         coins: SuiMoveObject[],
+        gasCoins: SuiMoveObject[],
         amount: bigint,
         validator: SuiAddress
     ): Promise<SuiExecuteTransactionResponse> {
-        const coin = await Coin.requestSuiCoinWithExactAmount(
-            signer,
-            coins,
-            amount
+        // sort to get the smallest one for gas
+        const sortedGasCoins = CoinAPI.sortByBalance(gasCoins);
+        const gasCoin = CoinAPI.selectCoinWithBalanceGreaterThanOrEqual(
+            sortedGasCoins,
+            BigInt(DEFAULT_GAS_BUDGET_FOR_STAKE)
         );
+        if (!gasCoin) {
+            throw new Error(
+                'Insufficient funds, not enough funds to cover for gas fee.'
+            );
+        }
+        if (!coins.length) {
+            throw new Error('Insufficient funds, no coins found.');
+        }
+        const isSui = CoinAPI.getCoinTypeArg(coins[0]) === SUI_TYPE_ARG;
+        const stakeCoins =
+            CoinAPI.selectCoinSetWithCombinedBalanceGreaterThanOrEqual(
+                coins,
+                amount,
+                isSui ? [CoinAPI.getID(gasCoin)] : undefined
+            ).map(CoinAPI.getID);
+        if (!stakeCoins.length) {
+            if (stakeCoins.length === 1 && isSui) {
+                throw new Error(
```

Comment:
I thought for form validation covers this?

Reply:
Yeah the validation I think has a check for this but since this can be called from anywhere I just added the checks